src/AsyncProducers.cpp

```diff
@@ -369,11 +369,22 @@ class ForkAsyncProducers : public IRMutator2 {
 class InitializeSemaphores : public IRMutator2 {
     using IRMutator2::visit;
 
+    const Type sema_type = type_of<halide_semaphore_t *>();
+
     Stmt visit(const LetStmt *op) {
-        Stmt body = mutate(op->body);
-        if (op->value.type() == type_of<halide_semaphore_t *>()) {
-            vector<pair<string, Expr>> lets;
+        vector<const LetStmt *> frames;
+
+        // Find first op that is of sema_type
+        while (op && op->value.type() != sema_type) {
+            frames.push_back(op);
+            op = op->body.as<LetStmt>();
+        }
+
+        Stmt body;
+        if (op) {
+            body = mutate(op->body);
             // Peel off any enclosing lets
+            vector<pair<string, Expr>> lets;
```

Comment:
I was deeply confused by the two vectors apparently doing the same thing until I realized the first was letstmts and the second was lets surrounding the value of the innermost letstmt. Please adjust comment to "Peel off any enclosing let expressions from the value" or similar.